sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) w.vlenterie
+ *
+ * This file is part of paparazzi
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/calibration/mag_calib_ukf.c"
+ * @author w.vlenterie
+ * Calibrate the magnetometer using an unscented kalman filter
+ * For more information please visit the following links:
+ *   - https://github.com/sfwa/trical
+ *   - http://au.tono.my/log/20131213-trical-magnetometer-calibration.html
+ *   - http://www.acsu.buffalo.edu/~johnc/mag_cal05.pdf
+ */
+
+#include <stdio.h>
+#include <error.h>
+#include <stdbool.h>
+#include <time.h>
+
+#include "modules/calibration/mag_calib_ukf.h"
+#include "subsystems/ahrs/ahrs_magnetic_field_model.h"
+#include "math/pprz_algebra_double.h"
+#include "TRICAL.h"
+#include "state.h"
+#include "subsystems/imu/imu_bebop.h"
+#include "modules/geo_mag/geo_mag.h"
+#include "generated/airframe.h"
+#include "subsystems/gps.h"
+
+
+#define PRINT(string,...) fprintf(stderr, "[CALIB_UKF->%s()] " string,__FUNCTION__ , ##__VA_ARGS__)
+
+#if !defined MAG_CALIB_UKF_GEO_MAG_TIMEOUT
+#define MAG_CALIB_UKF_GEO_MAG_TIMEOUT 0
+#else
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_GEO_MAG_TIMEOUT)
+#endif
+
+#if !defined MAG_CALIB_UKF_NORM
+#define MAG_CALIB_UKF_NORM 1.0f
+#else
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NORM)
+#endif
+
+#if !defined MAG_CALIB_UKF_NOISE_RMS
+#define MAG_CALIB_UKF_NOISE_RMS 2e-1f
+#else
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NOISE_RMS)
+#endif
+
+#if !defined MAG_CALIB_UKF_HOTSTART
+#define MAG_CALIB_UKF_HOTSTART TRUE
+#else
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART)
+#endif
+
+#if !defined MAG_CALIB_UKF_HOTSTART_SAVE_FILE
+#define MAG_CALIB_UKF_HOTSTART_SAVE_FILE /data/ftp/internal_000/mag_ukf_calib.txt
+#else
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART_SAVE_FILE)
+#endif
+
+#if !defined MAG_CALIB_UKF_VERBOSE
+#define VERBOSE_PRINT(...)
+#elif MAG_CALIB_UKF_VERBOSE == TRUE
+#define VERBOSE_PRINT PRINT
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_VERBOSE)
+#else
+#define VERBOSE_PRINT(...)
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_VERBOSE)
+#endif
+
+#if USE_MAGNETOMETER
+TRICAL_instance_t mag_calib;
+
+time_t mag_field_from_geo_mag;
+bool   update_geo_mag_field = true;
+struct FloatVect3 H = { .x = MAG_CALIB_UKF_NORM, .y = 0.0f, .z =  0.0f};
+#endif
+
+#if USE_MAGNETOMETER && MAG_CALIB_UKF_HOTSTART
+static FILE* fp;
+char hotstart_file_name[512];
+#endif
+
+void mag_calib_ukf_init(struct Imu *_imu) {
```

Comment:
Would a warning  to remove calibration from airframe file when `IMU_MAG_X_NEUTRAL`, `IMU_MAG_Y_NEUTRAL` or `IMU_MAG_Z_NEUTRAL` is defined suffice?

I don't think there is any real disadvantage to defining them, except that they are not necessary, since the UKF will correct for any wrongly entered NEUTRAL value.
Changing these values in your airframe file between flights however will upset the hot-start functionality.